### PR TITLE
db/nested-layouts-serialization

### PR DESCRIPTION
**Pull request by pawellysy**
Implemented nested layout serialization.

### CODE REVIEW

**Comment by pawellysy** [comment]
TODO:

- [ ] Test

File: ts/Dashboards/Layout/Cell.ts [hunk 1]
```diff
@@ -599,6 +648,10 @@ namespace Cell {
          * To create a nested layout, cell might contain an additional layouts.
          **/
         layout?: LayoutType.Options;
+        /**
+         * To create nested layout from JSON config.
```
Comment (karolkolodziej) [inline ts/Dashboards/Layout/Cell.ts:652]: Please mark as `internal` just in case

**Review by karolkolodziej — APPROVED** [review]

File: test/typescript-karma/Dashboards/Layout/tests.js [hunk 2]
```diff
@@ -182,3 +182,77 @@ test('Components and rows in layout with set height', function (assert) {
...
+    const exportedRows = layoutToExport.rows;
+    const exportedRowsLength = layoutToExport.rows.length;
+    const exportedCellsLength = exportedRows[0].cells.length;
+    const numberOfMountedComponents = board.mountedComponents.length;
+    layoutToExport.exportLocal();
+    layoutToExport.destroy();
+    const importedLayout = board.importLayoutLocal(exportedLayoutId);
+
+
+    assert.equal(importedLayout.rows.length, exportedRowsLength, 'The imported layout has an equal number of rows as exported one.')
```
Comment (sebastianbochan) [inline test/typescript-karma/Dashboards/Layout/tests.js:254]: You can use a dedicated assertion for checking the length.

`importedLayout.rows.length.should('have.length', exportedRowsLength)`

File: ts/Dashboards/Layout/Layout.ts [hunk 3]
```diff
@@ -282,6 +285,12 @@ class Layout extends GUIElement {
     public destroy(): void {
         const layout = this;
 
+        for (let i = layout.board.layouts.length - 1; i >= 0; i--) {
+            if (layout.board.layouts[i] === layout) {
```
Comment (sebastianbochan) [inline ts/Dashboards/Layout/Layout.ts:289]: Can we compare layouts by id? It will save us comparing the whole structure of layouts.

**Review by sebastianbochan — CHANGES_REQUESTED** [review]

File: ts/Dashboards/Layout/Layout.ts (hunk 3 above)
Comment (pawellysy) [reply ts/Dashboards/Layout/Layout.ts:289]: When you are comparing the objects in JS it is done by checking the reference of the object, so IMO it is safer to check if this is the same object.

File: test/typescript-karma/Dashboards/Layout/tests.js (hunk 2 above)
Comment (pawellysy) [reply test/typescript-karma/Dashboards/Layout/tests.js:254]: it is not cypress but karma, there is no such assertion here, or am I wrong?

Comment (sebastianbochan) [reply test/typescript-karma/Dashboards/Layout/tests.js:254]: Sorry, my mistake. I was pretty sure that is cypress test.

**Review by sebastianbochan — APPROVED** [review]

**Pull request merged**